domain/src/main/java/org/oppia/android/domain/topic/TopicListController.kt

```diff
@@ -486,38 +471,80 @@ class TopicListController @Inject constructor(
+    )
     // Suggest prerequisite topic user needs to learn after completing any of the topics.
-    for (i in 0 until topicIdList.size) {
-      val topicInProgressFound = topicProgressList.any { it.topicId == topicIdJsonArray[i] }
-      if (!topicInProgressFound) {
-        val dependentListOfTopics =
-          retrieveTopicDependencies(topicIdList[i])
-        for (j in dependentListOfTopics.indices) {
-          topicProgressList.mapIndexed { index, topicProgress ->
-            if (topicProgress.topicId == dependentListOfTopics[j]) {
-              val topic = topicController.retrieveTopic(topicProgressList[index].topicId)
-              if (checkIfAtLeastOneStoryIsCompleted(topicProgressList[index], topic)) {
-                val recommendedTopicAlreadyExist =
-                  recommendedStories.any { it.topicId == topicIdList[i] }
-                if (!recommendedTopicAlreadyExist) {
-                  val recommendedStoriesIdFromAssets =
-                    createRecommendedStoryFromAssets(topicIdList[i])
-                  if (recommendedStoriesIdFromAssets != null) {
-                    recommendedStories.add(recommendedStoriesIdFromAssets)
-                  }
-                }
-              }
-            }
-          }
+    for (topicId in unstartedTopicIdList) {
+      // All of the topic's prerequisites must be completed before it can be suggested.
+      val dependentTopicIds = topicDependencyMap[topicId]
+      if (topicId !in impliedFinishedTopicIds && startedTopicIds.containsAll(dependentTopicIds!!)) {
+        createRecommendedStoryFromAssets(topicId)?.let {
+          recommendedStories.add(it)
         }
       }
     }
     return recommendedStories
   }
 
+  private fun topicHasAtLeastOnStoryCompleted(it: TopicProgress): Boolean {
+    val topic = topicController.retrieveTopic(it.topicId)
+    it.storyProgressMap.values.forEach { storyProgress ->
+      val storyId = storyProgress.storyId
+      val story = topicController.retrieveStory(topic.topicId, storyId)
+      val isStoryCompleted = checkIfStoryIsCompleted(storyProgress, story)
+      if (isStoryCompleted) {
+        return true
+      }
+    }
+    return false
+  }
+
+  /** Returns the list of topic IDs that are completed or can be implied completed based on actually completed topics. */
+  private fun computeImpliedCompletedDependencies(
+    fullyCompletedTopicIds: List<String>,
+    topicDependencyMap: Map<String, Set<String>>
+  ): Set<String> {
+    // For each completed topic ID, compute the transitive closure of all of its dependencies & then combine them into a single list with the actual completed topic IDs. The returned list is a list of either completed or assumed completed topics which will eliminate potential recommendations.
+    return (
+      fullyCompletedTopicIds.flatMap { topicId ->
+        computeTransitiveDependencyClosure(topicId, topicDependencyMap)
+      } + fullyCompletedTopicIds
```

Comment:
Suggest moving the flatMap bit to a val to clean this up a bit & make it more readable.

Reply:
Done.